crates/bevy_ecs/src/query/state.rs

```diff
@@ -0,0 +1,291 @@
+            self.matched_archetype_ids.push(archetype.id());
+            if !self.matched_tables.contains(table_index) {
+                self.matched_tables.grow(table_index + 1);
+                self.matched_tables.set(table_index, true);
+                self.matched_table_ids.push(archetype.table_id());
+            }
+        }
+    }
+
+    #[inline]
+    pub fn get<'w>(
+        &mut self,
+        world: &'w World,
+        entity: Entity,
+    ) -> Result<<Q::Fetch as Fetch<'w>>::Item, QueryEntityError>
+    where
+        Q::Fetch: ReadOnlyFetch,
+    {
+        // SAFE: query is read only
+        unsafe { self.get_unchecked(world, entity) }
+    }
+
+    #[inline]
+    pub fn get_mut<'w>(
+        &mut self,
+        world: &'w mut World,
+        entity: Entity,
+    ) -> Result<<Q::Fetch as Fetch<'w>>::Item, QueryEntityError> {
+        // SAFE: query has unique world access
+        unsafe { self.get_unchecked(world, entity) }
+    }
+
+    /// # Safety
+    /// This does not check for mutable query correctness. To be safe, make sure mutable queries
+    /// have unique access to the components they query.
+    #[inline]
+    pub unsafe fn get_unchecked<'w>(
+        &mut self,
+        world: &'w World,
+        entity: Entity,
+    ) -> Result<<Q::Fetch as Fetch<'w>>::Item, QueryEntityError> {
+        self.update_archetypes(world);
+        self.get_unchecked_manual(world, entity)
+    }
+
+    /// # Safety
+    /// This does not check for mutable query correctness. To be safe, make sure mutable queries
+    /// have unique access to the components they query.
+    pub unsafe fn get_unchecked_manual<'w>(
+        &self,
+        world: &'w World,
+        entity: Entity,
+    ) -> Result<<Q::Fetch as Fetch<'w>>::Item, QueryEntityError> {
+        let location = world
+            .entities
+            .get(entity)
+            .ok_or(QueryEntityError::NoSuchEntity)?;
+        if !self
+            .matched_archetypes
+            .contains(location.archetype_id.index())
+        {
+            return Err(QueryEntityError::QueryDoesNotMatch);
+        }
+        // SAFE: live entities always exist in an archetype
+        let archetype = world.archetypes.get_unchecked(location.archetype_id);
+        let mut fetch = <Q::Fetch as Fetch>::init(world, &self.fetch_state);
+        let mut filter = <F::Fetch as Fetch>::init(world, &self.filter_state);
+
+        fetch.set_archetype(&self.fetch_state, archetype, &world.storages().tables);
+        filter.set_archetype(&self.filter_state, archetype, &world.storages().tables);
+        if filter.archetype_filter_fetch(location.index) {
+            Ok(fetch.archetype_fetch(location.index))
+        } else {
+            Err(QueryEntityError::QueryDoesNotMatch)
+        }
+    }
+
+    #[inline]
+    pub fn iter<'w, 's>(&'s mut self, world: &'w World) -> QueryIter<'w, 's, Q, F>
+    where
+        Q::Fetch: ReadOnlyFetch,
+    {
+        self.update_archetypes(world);
```

Comment:
`update_archetypes` is already called in `iter_unchecked` (this also applies to `iter_mut`)